File: server_epoll.c
#include <stdio.h>
#include <time.h>
#include <string.h>
#include <errno.h>
#include <sys/epoll.h>
#include "socket_hdr.h"
#include "base64.h"

#define IP_ADDRESS "0.0.0.0"
#define PORT		20000
#define BUF_SIZE	64
#define MAX_EPOLL_SIZE 1024



int main() {
    int svr_sock;                //套接字句柄
    int clifd;
    int result;
    int pos;
    char buf[BUF_SIZE];
    char* pB64;
    struct sockaddr_in addr;
    int timeout = 5000;
    int yes = 1;
    struct epoll_event event;
    struct epoll_event* events;
    int efd;
    
	if (init_socket() == -1) {
		printf("init socket error");
		return -1;
	}

	if ((svr_sock = socket(AF_INET, SOCK_STREAM, 0)) == 0){
		printf("create socket failed::%d\n", errno);
		return -1;
	}
    
    efd = epoll_create1(0);
    if (efd == -1){
        printf("create epoll error");
		return -1;
    }
    event.events = EPOLLIN;
    event.data.fd = svr_sock;
    if (epoll_ctl(efd, EPOLL_CTL_ADD, svr_sock, &event) == -1) {
        perror("epoll_ctl: listen_sock");
        return -1;
    }
    events = malloc(MAX_EPOLL_SIZE * sizeof(event));

    setsockopt(svr_sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));
    
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(PORT);
    if(bind(svr_sock, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
        printf("bind socket error\n");
        return -1;
    }
    
    if(listen(svr_sock, 5) == -1) {
        printf("listen socket error\n");
        return -1;
    }
    
    for(;;){
        // result = poll(pollfds, MAX_POLL_SIZE, timeout);
        result = epoll_wait(efd, events, MAX_EPOLL_SIZE, timeout);
        switch(result){        //开始监控
            case -1:                    //函数调用出错
                printf("epoll error \r\n");
                break;
            case 0:
                printf("time out \r\n");
                break;
            default:                    //得到数据返回
                for(int i = 0;i < result;i++) {
                    if(events[i].data.fd == svr_sock) {
                        clifd = accept(svr_sock, NULL, NULL);
                        printf("client accpet\n");
                        event.events = EPOLLIN | EPOLLET;
                        event.data.fd = clifd;
                       if (epoll_ctl(efd, EPOLL_CTL_ADD, clifd, &event) == -1) {
                           printf("epoll_ctl: conn_sock");
                           break;
                       }
                    } else if(events[i].events & EPOLLIN){
                        int n = recv(events[i].data.fd, buf, 64, 0);
                        printf("n ====>> %d\n", n);
                        if(n == 0) {
                            // close
                            close(events[i].data.fd);
                       } else if(n == -1) {
                           // error
                            close(events[i].data.fd);
                        } else {
                            buf[n] = 0;
                            pB64 = base64_encode(buf);
				            printf("%s <> %s\n", pB64, buf);
                            sprintf(buf, "%s", base64_decode(pB64));
                            send(events[i].data.fd, buf, strlen(buf), 0);
                        }
                    }
                }
                break;
        }
    }

    free(events);
    close(svr_sock);
    return 0;
}

File: server_epoll.py
import socket
import select
import sys
import queue

def run_server():
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setblocking(False)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server_address = ('0.0.0.0', 20002)
    server.bind(server_address)
    server.listen(5)
    message_queues = {}

    # 常用的标识  代表你想检查的事件类型
    READ_ONLY = select.EPOLLIN | select.EPOLLPRI | select.EPOLLHUP | select.EPOLLERR
    READ_WRITE = READ_ONLY | select.EPOLLOUT

    TIMEOUT = 1000
    epoller = select.epoll()
    epoller.register(server.fileno(), READ_ONLY)
    fd_to_socket = {server.fileno():server}

    while True:
        print("等待活动连接......")
        #轮询注册的事件集合，返回值为[(文件句柄，对应的事件)，(...),....]
        events = epoller.poll(TIMEOUT)
        if not events:
            print("epoll超时无活动连接，重新轮询......")
            continue
        print("有" , len(events), "个新事件，开始处理......")

        for fd, flag in events:
            s = fd_to_socket[fd]
            if flag & (select.EPOLLIN | select.EPOLLPRI): # 有数据可以读取
                if s is server: # 表示有新的连接
                    # 可以读取数据
                    connection, client_address = s.accept()
                    print(sys.stderr, '新的连接来自:', client_address)
                    connection.setblocking(False)
                    fd_to_socket[connection.fileno()] = connection # 往fd字典中添加一个新的 文件描述符
                    epoller.register(connection, READ_ONLY)
                    message_queues[connection] = queue.Queue() # 为了防止等待客户端发来数据期间发生阻塞，分配一个队列用于保存数据
                else: # 表示客户端传来了消息
                    data = s.recv(1024)
                    if data: # 表明数据接受成功
                        print(sys.stderr, '接受数据 "%s" 来自 %s' % (data, s.getpeername()))
                        message_queues[s].put(data)
                        # 修改一个已经存在的fd，修改事件为写。这里表示服务器向客户端要发送数据
                        epoller.modify(s, READ_WRITE)
                    else:
                        # 如果没有接受到数据，表示要断开连接
                        print(sys.stderr, '关闭', client_address, '并未读取到数据')
                        # 停止监听连接上的输入
                        epoller.unregister(s)
                        s.close()
                        # 将此链接从队列中删除
                        del message_queues[s]

            elif flag & select.EPOLLHUP:
                print(sys.stderr, '关闭', client_address, '收到HUP后')
                epoller.unregister(s)
                s.close()

            elif flag & select.EPOLLOUT:
                try:
                    next_msg = message_queues[s].get_nowait()
                except queue.Empty:
                    print(sys.stderr, '队列', s.getpeername(), '为空')
                    epoller.modify(s, READ_ONLY)
                else:
                    print(sys.stderr, '发送 "%s" 到 %s' % (next_msg, s.getpeername()))
                    s.send(next_msg)

            elif flag & select.EPOLLERR:
                print(sys.stderr, '异常信息:', s.getpeername())
                epoller.unregister(s)
                s.close()
                del message_queues[s]

if __name__ == "__main__":
    run_server()

